include_core/OMR/TypeTraits.hpp

```diff
@@ -0,0 +1,140 @@
+/*******************************************************************************
+ *  Copyright (c) 2018, 2018 IBM and others
+ *
+ *  This program and the accompanying materials are made available under
+ *  the terms of the Eclipse Public License 2.0 which accompanies this
+ *  distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ *  or the Apache License, Version 2.0 which accompanies this distribution and
+ *  is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ *  This Source Code may also be made available under the following
+ *  Secondary Licenses when the conditions for such availability set
+ *  forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ *  General Public License, version 2 with the GNU Classpath
+ *  Exception [1] and GNU General Public License, version 2 with the
+ *  OpenJDK Assembly Exception [2].
+ *
+ *  [1] https://www.gnu.org/software/classpath/license.html
+ *  [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ *  SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#if !defined(OMR_TYPETRAITS_HPP_)
+#define OMR_TYPETRAITS_HPP_
+
+namespace OMR
+{
+
+///
+/// Basic constants
+///
+
+/// Define an integral constant.
+template <typename T, T value>
+struct IntegralConstant
+{
+	typedef T ValueType;
+
+	static const ValueType VALUE = value;
+
+	ValueType operator()() const { return VALUE; }
+};
+
+template <typename T, T value>
+const T IntegralConstant<T, value>::VALUE;
+
+template <bool value>
+struct BoolConstant : IntegralConstant<bool, value> {};
+
+struct FalseConstant : BoolConstant<false> {};
+
+struct TrueConstant : BoolConstant<true> {};
+
+template <typename T>
+struct TypeAlias
+{
+	typedef T Type;
+};
+
+///
+/// Type modifications: add or remove qualifiers or type modifiers
+///
+
+/// RemoveConst<T> is T without any const qualifiers.
+template <typename T>
+struct RemoveConst : TypeAlias<T> {};
+
+template <typename T>
+struct RemoveConst<const T> : TypeAlias<T> {};
+
+/// RemoveVolatile<T>::Type is T without volatile qualifiers.
+template <typename T>
+struct RemoveVolatile : TypeAlias<T> {};
+
+template <typename T>
+struct RemoveVolatile<volatile T> : TypeAlias<T> {};
+
+/// RemoveCv<T>::Type is T without const/volatile qualifiers.
+template <typename T>
+struct RemoveCv : RemoveVolatile<typename RemoveConst<T>::Type> {};
+
+/// RemoveReference<T>::Type is T as a non-reference.
+template <typename T>
+struct RemoveReference : TypeAlias<T> {};
+
+template <typename T>
+struct RemoveReference<T&> : RemoveReference<T> {};
+
+// TODO: Handle RValue references in typetraits
```

Comment:
So, I don't know how other's feel about TODO comments, but I generally prefer GitHub issues because they are easier to track (I know we're actively trying to eliminate TODOs from the compiler component).

Reply:
To me it depends on the TODO and how long I think it will live in the code base.  When people have the code checked out (maybe even from another project like OpenJ9) knowing that there are GitHub Issues associated with the code you are looking at is very difficult. The TODO in the code allows users to know there is outstanding work. I personally do not like the idea of having hundreds if not thousands of GitHub Issues that may sit forgotten for years until we decide to close them when a simple TODO would suffice. I think committers can use their best judgement to ask for an Issue or leave the TODO.

Reply:
I see your point, although I think TODO's are just as likely to be forgotten and deleted :grin: .

In this case, I think I'm ok with the TODOs because it's very clear what the task to be done is. At most, I expect any discussion to be about what support/use for rvalue references OMR has. We can cross that bridge when we get to it.